### DISCUSSION TITLE
The bias of official implementation in GATConv and GATv2Conv

### DISCUSSION THREAD

**Discussion by hatute** [pyg-team/pytorch_geometric#6163] [Q&A] [answered]
After taking a deep look at the source code of GATConv and GATv2Conv, although the difference between them is limited, I have a common concern to the implementation of these two conv layers.

### Paper 1 
From the original paper of GATConv, "GRAPH ATTENTION NETWORKS" by Velickovic et al. We can see that double vertical lines denote for "Concatenation".

![Screenshot_2022-12-06_21-59-23](https://user-images.githubusercontent.com/15870636/206077554-28aea7b8-bbdf-4478-8a17-2e0508d45a27.png)

### Paper 2
From the original paper of GATv2Conv, "HOW ATTENTIVE ARE GRAPH ATTENTION NETWORKS?" by Brody et al. We can see that they means "Concatenation" as well.

![Screenshot_2022-12-06_22-00-02](https://user-images.githubusercontent.com/15870636/206077779-90aac065-091a-44b2-9cf7-db0f2bf41812.png)

### Paper 1
In the code of [GATConv](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/conv/gat_conv.py),
The first line inside method edge_update() is 
```
# Given edge-level attention coefficients for source and target nodes,
# we simply need to sum them up to "emulate" concatenation:
alpha = alpha_j if alpha_i is None else alpha_j + alpha_i
```

### Paper 2
In the code of [GATv2Conv](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/conv/gatv2_conv.py),
The first line inside method message() is 
```
 x = x_i + x_j
```

## Questions
So in essence, both of the methods replace the concatenation with the addition. Mathematically, these two operations are not equivalent. Could you guys give me more information about the correctness and precision behind this? 
If I miss something, please let me know ;) Thank you.

**Comment by wsad1** [comment, marked as answer]
`a.[k||h] = a.k + a.h`. The implementation is using this equivalence to convert concatentation and then dot product TO dot product and addition. 
So the formula from the paper and the code in pyg achieve the same thing.

**Reply by rusty1s** [reply]
Indeed. We first apply the transformation and then add the results together, which is equivalent to first concatenating the embeddings and then applying the transformation. The former is more efficient since it avoids creating the concatenated intermediate embedding.

**Reply by hatute** [reply]
Thank you guys for the explanation. It makes perfect sense.
